src/Oscoin/Data/Core.hs

```diff
@@ -0,0 +1,502 @@
+alter
+    :: (Maybe (StateVal c) -> Maybe (StateVal c))
+    -> StateKey
+    -> WorldState c
+    -> WorldState c
+alter f k ws =
+    case f (WorldState.lookup k ws) of
+        Just v  -> WorldState.insert k v ws
+        Nothing -> WorldState.delete k ws
+
+-------------------------------------------------------------------------------
+-- Core types
+-------------------------------------------------------------------------------
+
+-- | A balance of oscoin in the smallest denomination.
+type Balance = Word64
+
+-- | An account nonce to prevent replay attacks.
+type Nonce = Word64
+
+-- | An account which holds a balance.
+data Account c = Account
+    { accountAddr    :: Address c
+    -- ^ The account identifier.
+    , accountBalance :: Balance
+    -- ^ The oscoin balance.
+    , accountNonce   :: Nonce
+    -- ^ The nonce is equal to the number of transactions
+    -- made from this account.
+    }
+
+mkAccount :: Address c -> Account c
+mkAccount addr = Account
+    { accountAddr    = addr
+    , accountBalance = 0
+    , accountNonce   = 0
+    }
+
+-- | Increase the account balance.
+deposit :: Balance -> Account c -> Account c
+deposit bal acc = acc { accountBalance = accountBalance acc + bal }
+
+-- | Decrease the account balance.
+withdraw :: Balance -> Account c -> Account c
+withdraw bal acc = acc { accountBalance = accountBalance acc - bal }
+
+-- | Convert an 'Address' into a StateKey
+addressKey
+    :: ByteArrayAccess (Crypto.ShortHash c)
+    => CBOR.Serialise (Crypto.ShortHash c)
+    => Address c -> StateKey
+addressKey = serializeAddress
+
+-------------------------------------------------------------------------------
+
+-- | A number of blocks representing a long period of time.
+type Epoch = Natural
+
+-- | A number of epochs.
+type Epochs = Natural
+
+-- | A contribution signed-off by a maintainer.
+type Signoff c = Signed c (Contribution c)
+
+-------------------------------------------------------------------------------
+-- Checkpoints
+-------------------------------------------------------------------------------
+
+-- | A project checkpoint.
+data Checkpoint c = Checkpoint
+    { checkpointNumber        :: Natural
+    -- ^ The checkpoint number, starting from zero.
+    , checkpointStateHash     :: Crypto.Hash c
+    -- ^ The hash of the project state at this checkpoint.
+    , checkpointContributions :: [Contribution c]
+    -- ^ The new contributions since the last checkpoint.
+    , checkpointDependencies  :: [DependencyUpdate c]
+    -- ^ Updates to the dependencies since the last checkpoint.
+    }
+
+instance Eq (Checkpoint c) where
+    a == b = checkpointNumber a == checkpointNumber b
+
+instance Ord (Checkpoint c) where
+    a <= b = checkpointNumber a <= checkpointNumber b
+
+-- | A contribution to a project.
+data Contribution c = Contribution
+    { contribHash       :: Crypto.Hash c
+    -- ^ The hash of the off-chain contribution artefact.
+    , contribParentHash :: Maybe (Crypto.Hash c)
```

Comment:
What do you think about introducing `data ContributionArtifact` and changing this to `Crypto.Hashed ContriutionArtifact c`?

Reply:
Hmm I thought about this, but since these will come from the network, we won't have a `ContributionArtifact` to instantiate the type - what do you think?

Reply:
Come to think of it the tag will likely be only used here and we’ll unwrap it anyways. So let’s leave it like it is.